File: cucumber-expressions/go/parameter_type_registry.go
package cucumberexpressions

import (
	"fmt"
	"reflect"
	"regexp"
	"sort"
)

var INTEGER_REGEXPS = []*regexp.Regexp{
	regexp.MustCompile(`-?\d+`),
	regexp.MustCompile(`\d+`),
}
var FLOAT_REGEXPS = []*regexp.Regexp{
	regexp.MustCompile(`[-+]?\d*\.?\d+`),
}
var WORD_REGEXPS = []*regexp.Regexp{
	regexp.MustCompile(`[^\s]+`),
}
var STRING_REGEXPS = []*regexp.Regexp{
	regexp.MustCompile(`"([^"\\]*(\\.[^"\\]*)*)"|'([^'\\]*(\\.[^'\\]*)*)'`),
}
var ANONYMOUS_REGEXPS = `.*`

type ParameterTypeRegistry struct {
	parameterTypeByName    map[string]*ParameterType
	parameterTypesByRegexp map[string][]*ParameterType
	defaultTransformer     ParameterByTypeTransformer
}

func NewParameterTypeRegistry() *ParameterTypeRegistry {
	transformer := BuiltInParameterTransformer{}

	result := &ParameterTypeRegistry{
		parameterTypeByName:    map[string]*ParameterType{},
		parameterTypesByRegexp: map[string][]*ParameterType{},
		defaultTransformer:     transformer,
	}
	intParameterType, err := NewParameterType(
		"int",
		INTEGER_REGEXPS,
		"int",
		func(args ...*string) interface{} {
			i, err := transformer.Transform(*args[0], reflect.Int)
			if err != nil {
				panic(err)
			}
			return i
		},
		true,
		true,
		false,
	)
	if err != nil {
		panic(err)
	}
	result.DefineParameterType(intParameterType)
	floatParameterType, err := NewParameterType(
		"float",
		FLOAT_REGEXPS,
		"float",
		func(args ...*string) interface{} {
			f, err := transformer.Transform(*args[0], reflect.Float64)
			if err != nil {
				panic(err)
			}
			return f
		},
		true,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	result.DefineParameterType(floatParameterType)
	wordParameterType, err := NewParameterType(
		"word",
		WORD_REGEXPS,
		"string",
		func(args ...*string) interface{} {
			i, err := transformer.Transform(*args[0], reflect.String)
			if err != nil {
				panic(err)
			}
			return i
		},
		false,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	result.DefineParameterType(wordParameterType)
	stringParameterType, err := NewParameterType(
		"string",
		STRING_REGEXPS,
		"string",
		func(args ...*string) interface{} {
			if args[0] == nil && args[1] != nil {
				i, err := transformer.Transform(*args[1], reflect.String)
				if err != nil {
					panic(err)
				}
				return i
			}
			i, err := transformer.Transform(*args[0], reflect.String)
			if err != nil {
				panic(err)
			}
			return i
		},
		true,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	result.DefineParameterType(stringParameterType)

	anonymouseParameterType, err := createAnonymousParameterType(ANONYMOUS_REGEXPS)
	if err != nil {
		panic(err)
	}
	result.DefineParameterType(anonymouseParameterType)

	return result
}

func (p *ParameterTypeRegistry) ParameterTypes() []*ParameterType {
	result := make([]*ParameterType, len(p.parameterTypeByName))
	index := 0
	for _, parameterType := range p.parameterTypeByName {
		result[index] = parameterType
		index++
	}
	return result
}

func (p *ParameterTypeRegistry) LookupByTypeName(name string) *ParameterType {
	return p.parameterTypeByName[name]
}

func (p *ParameterTypeRegistry) LookupByRegexp(parameterTypeRegexp string, expressionRegexp string, text string) (*ParameterType, error) {
	parameterTypes, ok := p.parameterTypesByRegexp[parameterTypeRegexp]
	if !ok {
		return nil, nil
	}
	if len(parameterTypes) > 1 && !parameterTypes[0].PreferForRegexpMatch() {
		generatedExpressions := NewCucumberExpressionGenerator(p).GenerateExpressions(text)
		return nil, NewAmbiguousParameterTypeError(parameterTypeRegexp, expressionRegexp, parameterTypes, generatedExpressions)
	}
	return parameterTypes[0], nil
}

func (p *ParameterTypeRegistry) DefineParameterType(parameterType *ParameterType) error {
	if _, ok := p.parameterTypeByName[parameterType.Name()]; ok {
		if len(parameterType.Name()) == 0 {
			return fmt.Errorf("The anonymous parameter type has already been defined")
		}
		return fmt.Errorf("There is already a parameter type with name %s", parameterType.Name())
	}
	p.parameterTypeByName[parameterType.Name()] = parameterType
	for _, parameterTypeRegexp := range parameterType.Regexps() {
		if _, ok := p.parameterTypesByRegexp[parameterTypeRegexp.String()]; !ok {
			p.parameterTypesByRegexp[parameterTypeRegexp.String()] = []*ParameterType{}
		}
		parameterTypes := p.parameterTypesByRegexp[parameterTypeRegexp.String()]
		if len(parameterTypes) > 0 && parameterTypes[0].PreferForRegexpMatch() && parameterType.PreferForRegexpMatch() {
			return fmt.Errorf("There can only be one preferential parameter type per regexp. The regexp /%s/ is used for two preferential parameter types, {%s} and {%s}", parameterTypeRegexp.String(), parameterTypes[0].Name(), parameterType.Name())
		}
		parameterTypes = append(parameterTypes, parameterType)
		sort.Slice(parameterTypes, func(i int, j int) bool {
			return CompareParameterTypes(parameterTypes[i], parameterTypes[j]) <= 0
		})
		p.parameterTypesByRegexp[parameterTypeRegexp.String()] = parameterTypes
	}
	return nil
}

File: cucumber-expressions/javascript/src/ParameterTypeRegistry.ts
import ParameterType from './ParameterType'

import CucumberExpressionGenerator from './CucumberExpressionGenerator'
import {AmbiguousParameterTypeError, CucumberExpressionError} from './Errors'

export default class ParameterTypeRegistry {
  public static readonly INTEGER_REGEXPS = [/-?\d+/, /\d+/]
  public static readonly FLOAT_REGEXP = /(?=.*\d.*)[-+]?\d*(?:\.(?=\d.*))?\d*(?:\d+[E][+\-]?\d+)?/
  public static readonly WORD_REGEXP = /[^\s]+/
  public static readonly STRING_REGEXP = /"([^"\\]*(\\.[^"\\]*)*)"|'([^'\\]*(\\.[^'\\]*)*)'/
  public static readonly ANONYMOUS_REGEXP = /.*/

  private readonly parameterTypeByName = new Map<string, ParameterType<any>>()
  private readonly parameterTypesByRegexp = new Map<string, Array<ParameterType<any>>>()

  constructor() {
    this.defineParameterType(
      new ParameterType(
        'int',
        ParameterTypeRegistry.INTEGER_REGEXPS,
        Number,
        s => (s === undefined ? null : Number(s)),
        true,
        true
      )
    )
    this.defineParameterType(
      new ParameterType(
        'float',
        ParameterTypeRegistry.FLOAT_REGEXP,
        Number,
        s => (s === undefined ? null : parseFloat(s)),
        true,
        false
      )
    )
    this.defineParameterType(
      new ParameterType(
        'word',
        ParameterTypeRegistry.WORD_REGEXP,
        String,
        s => s,
        false,
        false
      )
    )
    this.defineParameterType(
      new ParameterType(
        'string',
        ParameterTypeRegistry.STRING_REGEXP,
        String,
        s => s.replace(/\\"/g, '"').replace(/\\'/g, "'"),
        true,
        false
      )
    )
    this.defineParameterType(
      new ParameterType(
        '',
        ParameterTypeRegistry.ANONYMOUS_REGEXP,
        String,
        s => s,
        false,
        true
      )
    )
  }

  get parameterTypes() {
    return this.parameterTypeByName.values()
  }

  public lookupByTypeName(typeName: string) {
    return this.parameterTypeByName.get(typeName)
  }

  public lookupByRegexp(
    parameterTypeRegexp: string,
    expressionRegexp: RegExp,
    text: string
  ): ParameterType<any> {
    const parameterTypes = this.parameterTypesByRegexp.get(parameterTypeRegexp)
    if (!parameterTypes) {
      return null
    }
    if (parameterTypes.length > 1 && !parameterTypes[0].preferForRegexpMatch) {
      // We don't do this check on insertion because we only want to restrict
      // ambiguiuty when we look up by Regexp. Users of CucumberExpression should
      // not be restricted.
      const generatedExpressions = new CucumberExpressionGenerator(
        this
      ).generateExpressions(text)
      throw AmbiguousParameterTypeError.forRegExp(
        parameterTypeRegexp,
        expressionRegexp,
        parameterTypes,
        generatedExpressions
      )
    }
    return parameterTypes[0]
  }

  public defineParameterType(parameterType: ParameterType<any>) {
    if (parameterType.name !== undefined) {
      if (this.parameterTypeByName.has(parameterType.name)) {
        if (parameterType.name.length === 0) {
          throw new Error(
            `The anonymous parameter type has already been defined`
          )
        } else {
          throw new Error(
            `There is already a parameter type with name ${parameterType.name}`
          )
        }
      }
      this.parameterTypeByName.set(parameterType.name, parameterType)
    }

    for (const parameterTypeRegexp of parameterType.regexpStrings) {
      if (!this.parameterTypesByRegexp.has(parameterTypeRegexp)) {
        this.parameterTypesByRegexp.set(parameterTypeRegexp, [])
      }
      const parameterTypes = this.parameterTypesByRegexp.get(
        parameterTypeRegexp
      )
      const existingParameterType = parameterTypes[0]
      if (
        parameterTypes.length > 0 &&
        existingParameterType.preferForRegexpMatch &&
        parameterType.preferForRegexpMatch
      ) {
        throw new CucumberExpressionError(
          'There can only be one preferential parameter type per regexp. ' +
            `The regexp /${parameterTypeRegexp}/ is used for two preferential parameter types, {${existingParameterType.name}} and {${parameterType.name}}`
        )
      }
      if (parameterTypes.indexOf(parameterType) === -1) {
        parameterTypes.push(parameterType)
        this.parameterTypesByRegexp.set(
          parameterTypeRegexp,
          parameterTypes.sort(ParameterType.compare)
        )
      }
    }
  }
}

module.exports = ParameterTypeRegistry
